src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py

```diff
@@ -0,0 +1,357 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This tests server certificate rotation support.
+
+Here we test various aspects of gRPC Python, and in some cases C-core
+by extension, support for server certificate rotation.
+
+* ServerSSLCertReloadTestWithClientAuth: test ability to rotate
+  server's SSL cert for use in future channels with clients while not
+  affecting any existing channel. The server requires client
+  authentication.
+
+* ServerSSLCertReloadTestWithoutClientAuth: like
+  ServerSSLCertReloadTestWithClientAuth except that the server does
+  not authenticate the client.
+
+* ServerSSLCertReloadTestCertConfigReuse: tests gRPC Python's ability
+  to deal with user's reuse of ServerCertificateConfig instances.
+
+"""
+
+import os
+import unittest
+from concurrent import futures
+
+import grpc
+from tests.unit import resources
+from tests.testing import _application_common
+from tests.testing import _server_application
+from tests.testing.proto import services_pb2_grpc
+
+ca_1_pem = resources.cert_hier_1_root_ca_cert()
+ca_2_pem = resources.cert_hier_2_root_ca_cert()
+
+client_key_1_pem = resources.cert_hier_1_client_1_key()
+client_chain_1_pem = '\n'.join([
+    resources.cert_hier_1_client_1_cert(),
+    resources.cert_hier_1_intermediate_ca_cert(),
+])
+
+client_key_2_pem = resources.cert_hier_2_client_1_key()
+client_chain_2_pem = '\n'.join([
+    resources.cert_hier_2_client_1_cert(),
+    resources.cert_hier_2_intermediate_ca_cert(),
+])
+
+server_key_1_pem = resources.cert_hier_1_server_1_key()
+server_cert_1_pem = '\n'.join([
+    resources.cert_hier_1_server_1_cert(),
+    resources.cert_hier_1_intermediate_ca_cert(),
+])
+
+server_key_2_pem = resources.cert_hier_2_server_1_key()
+server_cert_2_pem = '\n'.join([
+    resources.cert_hier_2_server_1_cert(),
+    resources.cert_hier_2_intermediate_ca_cert(),
+])
+
+
+class CertConfigFetcher(object):
+
+    def __init__(self, ca_pem, key_pem, cert_pem, switch_cert_on_client_num):
+        self.client_num = 0
+        self.switch_cert_on_client_num = switch_cert_on_client_num
+        # we simulate cb failures at 2 and 3
+        assert self.switch_cert_on_client_num > 3
+        self.cert_config = grpc.ssl_server_certificate_config(
+            [(key_pem, cert_pem)],
+            root_certificates=ca_pem,)
+
+    def __call__(self):
+        self.client_num += 1
+        if self.client_num in (2, 3):
+            raise ValueError(
+                'just an error for fun, should not affect the test')
+        if self.client_num != self.switch_cert_on_client_num:
+            return None
+        else:
+            # do the switch
+            ret = self.cert_config
+            self.cert_config = None
+            return ret
+
+
+class _ServerSSLCertReloadTest(object):
```

Comment:
Eww, yuck, a mixin!

[I'm no fan of implementation inheritance](https://www.youtube.com/watch?v=3MNVP9-hglc), but implementation inheritance is a better choice here than a mixin. Are you able to follow a pattern like
```py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import abc
import unittest

import six


class AbstractSuperClassTest(six.with_metaclass(abc.ABCMeta, unittest.TestCase)):

  def setUp(self):
    print('AbstractSuperClassTest.setUp!')

  def tearDown(self):
    print('AbstractSuperClassTest.tearDown!')


class AppleTest(AbstractSuperClassTest):

  def test_apple(self):
    print('AppleTest.test_apple!')


class BananaTest(AbstractSuperClassTest):

  def setUp(self):
    super(BananaTest, self).setUp()
    print('Banana.setUp!')

  def tearDown(self):
    print('Banana.tearDown!')
    super(BananaTest, self).tearDown()

  def test_banana(self):
    print('BananaTest.test_banana!')


if __name__ == '__main__':
  unittest.main(verbosity=2)
```
?

Reply:
will try!

Reply:
done, or at least getting closer?